File: docs/validators/node/validate.js
#!/usr/bin/env node
"use strict";
/**
 * Implementasi Node.js dari Validator Contract v1.
 *
 * Kesetaraan dengan implementasi referensi Python dibuktikan oleh conformance
 * suite yang sama (../conformance/) — bukan oleh kemiripan kode. Lihat
 * ../../spec/VALIDATOR_CONTRACT.md untuk perilaku yang wajib dipenuhi.
 * Tanpa dependensi eksternal: Node standar saja.
 *
 * Keyword yang didukung:
 *   type, required, properties, additionalProperties (bool),
 *   enum, const, pattern, minLength, maxLength,
 *   minItems, maxItems, uniqueItems, items,
 *   minimum, maximum, $ref (lokal: "#/...").
 *
 * Output: sama seperti kontrak §3 (mode --json = array kanonik per instance).
 * Exit code: 0 semua valid, 1 ada yang tidak, 2 argumen/berkas bermasalah.
 */

const fs = require("fs");

/** Escape satu segmen JSON Pointer (RFC 6901). */
function esc(token) {
  return token.replace(/~/g, "~0").replace(/\//g, "~1");
}

/** Serialisasi kanonik (kunci terurut) untuk perbandingan nilai. */
function stable(value) {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map(stable).join(",") + "]";
  }
  const keys = Object.keys(value).sort();
  return "{" + keys.map((k) => JSON.stringify(k) + ":" + stable(value[k])).join(",") + "}";
}

function typeMatches(value, expected) {
  switch (expected) {
    case "integer":
      return typeof value === "number" && Number.isInteger(value);
    case "number":
      return typeof value === "number";
    case "boolean":
      return typeof value === "boolean";
    case "string":
      return typeof value === "string";
    case "object":
      return value !== null && typeof value === "object" && !Array.isArray(value);
    case "array":
      return Array.isArray(value);
    case "null":
      return value === null;
    default:
      return true; // tipe tak dikenal: jangan gagalkan diam-diam
  }
}

function resolveRef(ref, root) {
  if (!ref.startsWith("#/")) {
    throw new Error(`hanya $ref lokal yang didukung, dapat: ${JSON.stringify(ref)}`);
  }
  let node = root;
  for (const raw of ref.slice(2).split("/")) {
    node = node[raw.replace(/~1/g, "/").replace(/~0/g, "~")];
  }
  return node;
}

function hasDuplicates(items) {
  const seen = new Set();
  for (const item of items) {
    const key = stable(item);
    if (seen.has(key)) return true;
    seen.add(key);
  }
  return false;
}

/**
 * Kembalikan daftar error kanonik: [{path, keyword}].
 * Kosong berarti valid. `path` adalah JSON Pointer (root = "").
 */
function validate(instance, schema, root, path) {
  if (root === undefined) root = schema;
  if (path === undefined) path = "";
  const errors = [];

  if (Object.prototype.hasOwnProperty.call(schema, "$ref")) {
    return validate(instance, resolveRef(schema["$ref"], root), root, path);
  }

  if (Object.prototype.hasOwnProperty.call(schema, "type")) {
    const list = Array.isArray(schema.type) ? schema.type : [schema.type];
    if (!list.some((t) => typeMatches(instance, t))) {
      return [{ path, keyword: "type" }]; // cek lanjutan tak bermakna
    }
  }

  if (Object.prototype.hasOwnProperty.call(schema, "const") && stable(instance) !== stable(schema.const)) {
    errors.push({ path, keyword: "const" });
  }

  if (Object.prototype.hasOwnProperty.call(schema, "enum") &&
      !schema.enum.some((e) => stable(e) === stable(instance))) {
    errors.push({ path, keyword: "enum" });
  }

  if (typeof instance === "string") {
    if ("minLength" in schema && instance.length < schema.minLength) {
      errors.push({ path, keyword: "minLength" });
    }
    if ("maxLength" in schema && instance.length > schema.maxLength) {
      errors.push({ path, keyword: "maxLength" });
    }
    if ("pattern" in schema && !new RegExp(schema.pattern).test(instance)) {
      errors.push({ path, keyword: "pattern" });
    }
  }

  if (typeof instance === "number") {
    if ("minimum" in schema && instance < schema.minimum) {
      errors.push({ path, keyword: "minimum" });
    }
    if ("maximum" in schema && instance > schema.maximum) {
      errors.push({ path, keyword: "maximum" });
    }
  }

  if (Array.isArray(instance)) {
    if ("minItems" in schema && instance.length < schema.minItems) {
      errors.push({ path, keyword: "minItems" });
    }
    if ("maxItems" in schema && instance.length > schema.maxItems) {
      errors.push({ path, keyword: "maxItems" });
    }
    if (schema.uniqueItems && hasDuplicates(instance)) {
      errors.push({ path, keyword: "uniqueItems" });
    }
    if ("items" in schema) {
      instance.forEach((item, i) => {
        errors.push(...validate(item, schema.items, root, `${path}/${i}`));
      });
    }
  }

  if (instance !== null && typeof instance === "object" && !Array.isArray(instance)) {
    for (const key of schema.required || []) {
      if (!Object.prototype.hasOwnProperty.call(instance, key)) {
        errors.push({ path: `${path}/${esc(key)}`, keyword: "required" });
      }
    }
    const props = schema.properties || {};
    for (const [key, value] of Object.entries(instance)) {
      const child = `${path}/${esc(key)}`;
      if (Object.prototype.hasOwnProperty.call(props, key)) {
        errors.push(...validate(value, props[key], root, child));
      } else if (schema.additionalProperties === false) {
        errors.push({ path: child, keyword: "additionalProperties" });
      }
    }
  }

  return errors;
}

function load(path) {
  return JSON.parse(fs.readFileSync(path, "utf8"));
}

function main(argv) {
  let args = argv.slice(2);
  let asJson = false;
  if (args[0] === "--json") {
    asJson = true;
    args = args.slice(1);
  }
  if (args.length < 2) {
    console.log("Pemakaian: validate.js [--json] <schema.json> <instance.json> ...");
    return 2;
  }

  let schema;
  try {
    schema = load(args[0]);
  } catch (exc) {
    console.error(`gagal memuat schema ${args[0]}: ${exc.message}`);
    return 2;
  }

  let allOk = true;
  const results = [];
  for (const instancePath of args.slice(1)) {
    let instance;
    try {
      instance = load(instancePath);
    } catch (exc) {
      console.error(`gagal memuat instance ${instancePath}: ${exc.message}`);
      return 2;
    }
    const errors = validate(instance, schema);
    const ok = errors.length === 0;
    allOk = allOk && ok;
    results.push({ instance: instancePath, valid: ok, errors });
  }

  if (asJson) {
    console.log(JSON.stringify(results, null, 2));
  } else {
    for (const res of results) {
      if (res.valid) {
        console.log(`VALID    ${res.instance}`);
      } else {
        console.log(`INVALID  ${res.instance}`);
        for (const err of res.errors) {
          console.log(`    - ${err.path || "(root)"}: gagal '${err.keyword}'`);
        }
      }
    }
  }
  return allOk ? 0 : 1;
}

module.exports = { validate };

if (require.main === module) {
  process.exit(main(process.argv));
}

File: docs/validators/python/validate.py
#!/usr/bin/env python3
"""Implementasi REFERENSI validator kontrak (subset JSON Schema draft-07).

Ini adalah SATU implementasi dari kontrak di ../../spec/VALIDATOR_CONTRACT.md.
Bahasa lain bebas mengganti file ini selama lolos conformance suite yang sama
(../conformance/). Tanpa dependensi eksternal: Python 3 standar saja.

Keyword yang didukung:
  type, required, properties, additionalProperties (bool),
  enum, const, pattern, minLength, maxLength,
  minItems, maxItems, uniqueItems, items,
  minimum, maximum, $ref (lokal: "#/...").

Output:
  - Mode manusia (default): baris VALID/INVALID + daftar error terbaca.
  - Mode kanonik (--json): array JSON, satu objek per instance:
      {"instance": <path>, "valid": <bool>,
       "errors": [{"path": <JSON Pointer>, "keyword": <str>}, ...]}
    Error dibandingkan sebagai HIMPUNAN (urutan tak penting).

Exit code: 0 semua valid, 1 ada yang tidak valid, 2 argumen/berkas bermasalah.
"""
from __future__ import annotations

import json
import re
import sys

_JSON_TYPES = {"object": dict, "array": list, "string": str, "null": type(None)}


def _esc(token: str) -> str:
    """Escape satu segmen JSON Pointer (RFC 6901)."""
    return token.replace("~", "~0").replace("/", "~1")


def _type_matches(value, expected: str) -> bool:
    if expected == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if expected == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if expected == "boolean":
        return isinstance(value, bool)
    py = _JSON_TYPES.get(expected)
    if py is None:
        return True  # tipe tak dikenal: jangan gagalkan diam-diam
    if py is str:
        return isinstance(value, str) and not isinstance(value, bool)
    return isinstance(value, py)


def _resolve_ref(ref: str, root: dict):
    if not ref.startswith("#/"):
        raise ValueError(f"hanya $ref lokal yang didukung, dapat: {ref!r}")
    node = root
    for token in ref[2:].split("/"):
        node = node[token.replace("~1", "/").replace("~0", "~")]
    return node


def validate(instance, schema: dict, root: dict | None = None, path: str = "") -> list[dict]:
    """Kembalikan daftar error kanonik: [{"path": <pointer>, "keyword": <str>}].

    Kosong berarti valid. `path` adalah JSON Pointer (root = "").
    """
    if root is None:
        root = schema
    errors: list[dict] = []

    if "$ref" in schema:
        return validate(instance, _resolve_ref(schema["$ref"], root), root, path)

    if "type" in schema:
        expected = schema["type"]
        expected_list = expected if isinstance(expected, list) else [expected]
        if not any(_type_matches(instance, t) for t in expected_list):
            return [{"path": path, "keyword": "type"}]  # cek lanjutan tak bermakna

    if "const" in schema and instance != schema["const"]:
        errors.append({"path": path, "keyword": "const"})

    if "enum" in schema and instance not in schema["enum"]:
        errors.append({"path": path, "keyword": "enum"})

    if isinstance(instance, str) and not isinstance(instance, bool):
        if "minLength" in schema and len(instance) < schema["minLength"]:
            errors.append({"path": path, "keyword": "minLength"})
        if "maxLength" in schema and len(instance) > schema["maxLength"]:
            errors.append({"path": path, "keyword": "maxLength"})
        if "pattern" in schema and re.search(schema["pattern"], instance) is None:
            errors.append({"path": path, "keyword": "pattern"})

    if isinstance(instance, (int, float)) and not isinstance(instance, bool):
        if "minimum" in schema and instance < schema["minimum"]:
            errors.append({"path": path, "keyword": "minimum"})
        if "maximum" in schema and instance > schema["maximum"]:
            errors.append({"path": path, "keyword": "maximum"})

    if isinstance(instance, list):
        if "minItems" in schema and len(instance) < schema["minItems"]:
            errors.append({"path": path, "keyword": "minItems"})
        if "maxItems" in schema and len(instance) > schema["maxItems"]:
            errors.append({"path": path, "keyword": "maxItems"})
        if schema.get("uniqueItems") and _has_duplicates(instance):
            errors.append({"path": path, "keyword": "uniqueItems"})
        if "items" in schema:
            for i, item in enumerate(instance):
                errors += validate(item, schema["items"], root, f"{path}/{i}")

    if isinstance(instance, dict):
        for key in schema.get("required", []):
            if key not in instance:
                errors.append({"path": f"{path}/{_esc(key)}", "keyword": "required"})
        props = schema.get("properties", {})
        for key, value in instance.items():
            child = f"{path}/{_esc(key)}"
            if key in props:
                errors += validate(value, props[key], root, child)
            elif schema.get("additionalProperties") is False:
                errors.append({"path": child, "keyword": "additionalProperties"})

    return errors


def _has_duplicates(items: list) -> bool:
    seen: set[str] = set()
    for item in items:
        key = json.dumps(item, sort_keys=True)
        if key in seen:
            return True
        seen.add(key)
    return False


def _format_human(path: str, keyword: str) -> str:
    where = path if path else "(root)"
    return f"{where}: gagal '{keyword}'"


def _load(path: str):
    with open(path, encoding="utf-8") as fh:
        return json.load(fh)


def main(argv: list[str]) -> int:
    args = argv[1:]
    as_json = False
    if args and args[0] == "--json":
        as_json = True
        args = args[1:]
    if len(args) < 2:
        print("Pemakaian: validate.py [--json] <schema.json> <instance.json> ...")
        return 2
    try:
        schema = _load(args[0])
    except (OSError, json.JSONDecodeError) as exc:
        print(f"gagal memuat schema {args[0]}: {exc}", file=sys.stderr)
        return 2

    all_ok = True
    results = []
    for instance_path in args[1:]:
        try:
            instance = _load(instance_path)
        except (OSError, json.JSONDecodeError) as exc:
            print(f"gagal memuat instance {instance_path}: {exc}", file=sys.stderr)
            return 2
        errors = validate(instance, schema)
        ok = not errors
        all_ok = all_ok and ok
        results.append({"instance": instance_path, "valid": ok, "errors": errors})

    if as_json:
        print(json.dumps(results, ensure_ascii=False, indent=2))
    else:
        for res in results:
            if res["valid"]:
                print(f"VALID    {res['instance']}")
            else:
                print(f"INVALID  {res['instance']}")
                for err in res["errors"]:
                    print(f"    - {_format_human(err['path'], err['keyword'])}")
    return 0 if all_ok else 1


if __name__ == "__main__":
    sys.exit(main(sys.argv))
